extension/background.js

```diff
@@ -0,0 +1,230 @@
+      browser.webRequest.onBeforeRequest.removeListener(this.webRequestHandler);
+    }
+
+    if (
+      browser.webRequest.onBeforeRedirect.hasListener(this.webRequestHandler)
+    ) {
+      console.debug("removing onBeforeRedirect listener");
+      browser.webRequest.onBeforeRedirect.removeListener(
+        this.webRequestHandler
+      );
+    }
+
+    // Retrieve the list of URL patterns to monitor with our listener.
+    //
+    // Note: search suggestions are system principal requests, so webRequest
+    // cannot intercept them.
+    const matchPatternsMap = await this.getMatchPatterns();
+    const patterns = Object.keys(matchPatternsMap);
+
+    if (patterns.length === 0) {
+      console.debug(
+        "not registering any listener because there is no URL to monitor"
+      );
+      return;
+    }
+
+    console.debug("registering onBeforeRequest listener");
+    browser.webRequest.onBeforeRequest.addListener(
+      this.webRequestHandler,
+      { urls: patterns },
+      ["blocking"]
+    );
+
+    // This one is needed in addition to `onBeforeRequest` because this
+    // extension might be registered before or after some extensions.
+    // Depending on that, our `onBeforeRequest` listener might not be called,
+    // which is why we also listen to `onBeforeRedirect`.
+    console.debug("registering onBeforeRedirect listener");
+    browser.webRequest.onBeforeRedirect.addListener(this.webRequestHandler, {
+      urls: patterns,
+    });
+  }
+
+  // `redirectUrl` is usually valid when the redirect has been detected via
+  // `onBeforeRedirect`, otherwise it's likely `undefined`.
+  webRequestHandler = async ({ requestId, url, redirectUrl }) => {
+    if (this.lastRequestIdReported === requestId) {
+      console.debug(`request ID '${requestId}' already reported, skipping...`);
+      return;
+    }
+
+    // When we detect a redirect, we read the request property, hoping to find
+    // an add-on ID corresponding to the add-on that initiated the redirect.
+    // It might not return anything when the redirect is a search server-side
+    // redirect but it can also be caused by an error.
+    let addonId = await browser.addonsSearchExperiment.getRequestProperty(
+      requestId,
+      "redirectedByExtension"
+    );
+
+    // When we did not find an add-on ID in the request property bag and the
+    // `redirectUrl` is both valid and different than the original URL. we
+    // likely detected a search server-side redirect.
+    const isServerSideRedirect =
+      !addonId && typeof redirectUrl !== "undefined" && url !== redirectUrl;
+
+    // Search server-side redirects are possible because an extension has
+    // registered a search engine, which is why we can (hopefully) retrieve the
+    // add-on ID.
+    if (isServerSideRedirect) {
+      const id = this.getAddonIdFromUrl(url);
+
+      // We shouldn't report built-in search engines.
+      if (!id.endsWith("@search.mozilla.org")) {
```

Comment:
Why?  Aside from that, non of our builtin engines have code so this isn't necessary.

Reply:
In a previous iteration, there was no check between `from` and `to` eTLDs so the extension was reporting the `google.com` search engine in some cases.